### DISCUSSION TITLE
How to set environment variables for test?

### DISCUSSION THREAD

**Discussion by feslima** [nvim-neotest/neotest#34] [Q&A] [answered]
When using `vim-ultest` I could use the `ultest_env` variable to set my environment variables for the test suite. I've tried digging around the project docs/source but couldn't find where I could do this here.

Is is possible to achieve this feature with `neotest`? If so, how?

Thanks in advance!

**Comment by rcarriga** [comment, marked as answer]
I hadn't actually allowed this using the run API but I've now added an `env` argument to the args table to supplement the environment :smile:

**Reply by feslima** [reply]
Nice. Thank you @rcarriga for the awesome plugin and support!

**Reply by feslima** [reply]
@rcarriga It's me again 😅. The problem now is I'm not able to debug tests that need env vars set. I'm using `pytest` and this is what I'm calling to debug a test: `require('neotest').run.run({ strategy = 'dap', env = _G.test_env })` where `_G.test_env` is a global variable with the env vars table that I parse from files.

This API change is valid for the python/pytest dap debug adapter? Because, when I debug the tests with the code above, the env vars aren't being loaded in debug mode.

**Reply by feslima** [reply]
I think that the env table should've been inserted in the `strategy_config` from [here](https://github.com/nvim-neotest/neotest-python/blob/de234d600e728f2350950355bb5c8d723ea7dff7/lua/neotest-python/init.lua#L119), because I remember that I've had to do something similar when using `ultest` + `dap` [debugging recipe](https://github.com/rcarriga/vim-ultest/wiki/Debugging-Recipes#pytestunittest). 

But I'm not sure if the env vars table is being set there in the `neotest-python` adapter.

**Reply by rcarriga** [reply]
Ah so this is not currently possible without users changing their config due to the behaviour of nvim-dap. I've opened an issue to see if that can change https://github.com/mfussenegger/nvim-dap/issues/587

**Reply by feslima** [reply]
Nevermind! I can simply pass the env table as to the adapter itself. It works now.

For anyone wondering how to do this, in the `neotest-python` config, pass your env table to the `dap` arg:
```lua
neotest.setup({
  ...
  require('neotest-python')({
    dap = { env = your_env_table },
    ...
  }),
  ...
})
```

**Reply by feslima** [reply]
Indeed, I do agree that passing the env per run options is less complex to config.

Thank you anyway.

**Reply by rcarriga** [reply]
Good to know that works :+1: As far as I know that's dependent on the adapter accepting the env argument so that works for debugpy but not necessarily for others

**Reply by jetaggart** [reply]
How would you do this for neotest-go? Sorry, this isn't super clear to me how we can set env vars per adapter.

**Comment by feslima** [comment]
@rcarriga Hey, it's me again! Quick question: the setup using `neotest.run.run({ env=my_env_table })` was changed? It's been a while since my last package update and, when I did update yesterday, it appears that `neotest.run.run({ env=my_env_table })` is ignoring the env table I set and sends nothing to the runner. Is there any procedure where I can debug this?

**Reply by feslima** [reply]
I've tried some crude debugging to see what values `spec` is receiving after [this line](https://github.com/nvim-neotest/neotest/blob/4f04d0a8bd9e2f517367fd436c612d45ae4e1a13/lua/neotest/client/runner.lua#L28). 

This is the `spec` value: 
![image](https://user-images.githubusercontent.com/26683783/178737437-4ebd0101-b40a-42a7-8c49-7dd924029094.png)

The `env` table set in `args` isn't set on `spec`. Is this intended to be like this?

**Reply by feslima** [reply]
I was able to make it work again using this:
```lua
if args.env then
  spec.env = args.env
end
```
after [this line](https://github.com/nvim-neotest/neotest/blob/4f04d0a8bd9e2f517367fd436c612d45ae4e1a13/lua/neotest/client/runner.lua#L65).

Hope it helps!

**Reply by jonatan-branting** [reply]
Shouldn't the adapter handle that? It receives `args` through `build_spec` on line 28 in your example.

Although support for env variables without explicit adapter support maybe isn't the worst idea, either.

**Reply by feslima** [reply]
@jonatan-branting I've tested earlier, this isn't working too. Since there no changes in the docs about the `env` being set on the runner, I thought this is still supported, even though he mentioned that its support is dependent on adapter and [this PR](https://github.com/mfussenegger/nvim-dap/issues/587) was closed without changes.

For now, this dirty "fix" is working

**Reply by rcarriga** [reply]
Hmm thought I replied to this but apparently not :facepalm: I fixed this a couple of days ago, you should see env vars working again!
